### **Exercici 1**

Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe.  
Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants.  
Justifica la teva elecció.

In [1]:
import pandas as pd
import numpy as np
df = pd.read_csv('DelayedFlights.csv')
pd.set_option('display.max_columns', 35)
df.head()

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0           0  2008      1           3          4   2003.0        1955   
1           1  2008      1           3          4    754.0         735   
2           2  2008      1           3          4    628.0         620   
3           4  2008      1           3          4   1829.0        1755   
4           5  2008      1           3          4   1940.0        1915   

   ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
0   2211.0        2225            WN        335  N712SW              128.0   
1   1002.0        1000            WN       3231  N772SW              128.0   
2    804.0         750            WN        448  N428WN               96.0   
3   1959.0        1925            WN       3920  N464WN               90.0   
4   2121.0        2110            WN        378  N726SW              101.0   

   CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  \
0           150.0    116.0     -14.0       8.0    IAD  TPA       810     4.0   
1           145.0    113.0       2.0      19.0    IAD  TPA       810     5.0   
2            90.0     76.0      14.0       8.0    IND  BWI       515     3.0   
3            90.0     77.0      34.0      34.0    IND  BWI       515     3.0   
4           115.0     87.0      11.0      25.0    IND  JAX       688     4.0   

   TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  WeatherDelay  \
0      8.0          0                N         0           NaN           NaN   
1     10.0          0                N         0           NaN           NaN   
2     17.0          0                N         0           NaN           NaN   
3     10.0          0                N         0           2.0           0.0   
4     10.0          0                N         0           NaN           NaN   

   NASDelay  SecurityDelay  LateAircraftDelay  
0       NaN            NaN                NaN  
1       NaN            NaN                NaN  
2       NaN            NaN                NaN  
3       0.0            0.0               32.0  
4       NaN            NaN                NaN

In [2]:
df = df.drop(columns = ['Unnamed: 0', 'Year'])
df = df.drop(columns = ['TailNum'])
df = df.drop(columns = ['FlightNum'])
df = df.drop(columns = ['DepTime', 'CRSDepTime', 'ArrTime', 'CRSArrTime'])

In [3]:
df.head(3)

Month  DayofMonth  DayOfWeek UniqueCarrier  ActualElapsedTime  \
0      1           3          4            WN              128.0   
1      1           3          4            WN              128.0   
2      1           3          4            WN               96.0   

   CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  \
0           150.0    116.0     -14.0       8.0    IAD  TPA       810     4.0   
1           145.0    113.0       2.0      19.0    IAD  TPA       810     5.0   
2            90.0     76.0      14.0       8.0    IND  BWI       515     3.0   

   TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  WeatherDelay  \
0      8.0          0                N         0           NaN           NaN   
1     10.0          0                N         0           NaN           NaN   
2     17.0          0                N         0           NaN           NaN   

   NASDelay  SecurityDelay  LateAircraftDelay  
0       NaN            NaN                NaN  
1       NaN            NaN                NaN  
2       NaN            NaN                NaN

Les dades fan referència als retards i cancel·lacions de vols als Estats Units al 2008.  
Les columnes reflecteixen les característiques dels vols per poder valorar les incidències (Data del vol, aeroport d'origen i destí, hores d'enlairament i aterratge, horaris previstos, temps de vol, retard a l'aire, companyia aèria..).
- Mitjançant .unique, he descartat les columnes que només tenen un valor (Unnamed o, Year).
- He descartat la columna 'TailNum', sembla raonable que les característiques tècniques de l'avió influeixin poc en un retard, en comparació amb altres variables. Part de la seva informació més rellevant seria la companyia a què pertany i aquesta informació està en una columna per si sola.
- He descartat també el número de vol. Encara que hi ha certes practiques habituals com adjudicar números parells als vols dirigits a l'Est i Nord (i imparells als contraris) o adjudicar tres digitos a la llarga distància, hi ha columnes amb les dades d'origen i destí o amb la distància del vol en milles .. A més no deixen de ser practiques habituals i no una dada fiable.
- Atès que les hores de sortida i arribada (tant reals com previstes) són locals i podrien existir canvis horaris i que ja existeixen les columnes amb la diferència real en minuts (tant reals com previstes), he decidit prescindir-ne també.

In [4]:
import copy
data = copy.copy(df)

In [5]:
data.to_csv('data_fly.csv', index = None )

### **Exercici 2**  

Fes un informe complet del dataset:  

Resumeix estadísticament el dataset i les columnes d’interès. Fes una anàlisi estadístic del que consideris rellevant.   
Troba quantes dades faltants hi ha per columna.  
Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).  
Fes una taula de les aerolínies amb més endarreriments acumulats.  
Quins són els vols més llargs? I els més endarrerits? Busca les rutes més llargues i les que acumulen més retards.  
Aporta allò que consideris rellevant.

In [6]:
import pandas as pd
import numpy as np
data = pd.read_csv('data_fly.csv')
pd.set_option('display.max_columns', 35)
data.head(5)

Month  DayofMonth  DayOfWeek UniqueCarrier  ActualElapsedTime  \
0      1           3          4            WN              128.0   
1      1           3          4            WN              128.0   
2      1           3          4            WN               96.0   
3      1           3          4            WN               90.0   
4      1           3          4            WN              101.0   

   CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  \
0           150.0    116.0     -14.0       8.0    IAD  TPA       810     4.0   
1           145.0    113.0       2.0      19.0    IAD  TPA       810     5.0   
2            90.0     76.0      14.0       8.0    IND  BWI       515     3.0   
3            90.0     77.0      34.0      34.0    IND  BWI       515     3.0   
4           115.0     87.0      11.0      25.0    IND  JAX       688     4.0   

   TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  WeatherDelay  \
0      8.0          0                N         0           NaN           NaN   
1     10.0          0                N         0           NaN           NaN   
2     17.0          0                N         0           NaN           NaN   
3     10.0          0                N         0           2.0           0.0   
4     10.0          0                N         0           NaN           NaN   

   NASDelay  SecurityDelay  LateAircraftDelay  
0       NaN            NaN                NaN  
1       NaN            NaN                NaN  
2       NaN            NaN                NaN  
3       0.0            0.0               32.0  
4       NaN            NaN                NaN

In [7]:
data.shape

(1936758, 22)

In [8]:
Comp = list(data['UniqueCarrier'].unique())
print('Num de companyies: ' + str(len(Comp)))
Aerop = list(data['Origin'].unique())
print('Num de aeroports: ' + str(len(Aerop)))
Causa = list(data['CancellationCode'].unique())
print('Num de causes de retard: ' + str(len(Causa)))

Num de companyies: 20
Num de aeroports: 303
Num de causes de retard: 4


En la tabla, aparecen 20 compañías aereas.  
Aparecen 303 Aeropuertos.  
Aparecen 4 causas de retraso.  

In [9]:
print(len(data['CarrierDelay']),len(data['WeatherDelay']),len(data['NASDelay']),len(data['SecurityDelay']),len(data['LateAircraftDelay']))  

1936758 1936758 1936758 1936758 1936758


In [10]:
data['CarrierDelay'] = data['CarrierDelay'].fillna(0)  
data['WeatherDelay'] = data['WeatherDelay'].fillna(0)  
data['NASDelay'] = data['NASDelay'].fillna(0)  
data['SecurityDelay'] = data['SecurityDelay'].fillna(0)  
data['LateAircraftDelay'] = data['LateAircraftDelay'].fillna(0)  
Ret_Comp = sum(data['CarrierDelay'])  
Ret_Meteo = sum(data['WeatherDelay'])  
Ret_Aerop = sum(data['NASDelay'])  
Ret_Seg = sum(data['SecurityDelay'])
# (Este de momento no lo uso porque es el retraso del vuelo anterior acumulado -creo-)
Ret_Ret = sum(data['LateAircraftDelay'])
Ret_Total = Ret_Comp + Ret_Meteo + Ret_Aerop + Ret_Seg  

Como me interesa saber el retraso acumulado por cada causa, he sustituido los valores NaN por 0 y he hecho la suma de toda la columna, sabiendo así cuánto acumula cada retraso a lo largo del año.  
Se puede apreciar que practicamente todo el retraso es causa de las companyies o els aeroports.

In [11]:
Ret_Comp_pc = round((Ret_Comp * 100) / Ret_Total, 2)
Ret_Meteo_pc = round((Ret_Meteo * 100) / Ret_Total, 2)
Ret_Aerop_pc = round((Ret_Aerop * 100) / Ret_Total, 2)
Ret_Seg_pc = round((Ret_Seg * 100) / Ret_Total, 2)

print('Retards per causa de la companyia: ' + str(Ret_Comp_pc) + ' %.')
print('Retards per causa del aeroport: ' + str(Ret_Aerop_pc) + ' %.')
print('Retards per causa de la meteorologia: ' + str(Ret_Meteo_pc) + ' %.')
print('Retards per causa de seguretat: ' + str(Ret_Seg_pc) + ' %.')

Retards per causa de la companyia: 50.48 %.
Retards per causa del aeroport: 39.54 %.
Retards per causa de la meteorologia: 9.75 %.
Retards per causa de seguretat: 0.24 %.


In [12]:
Cancel_Month = data.groupby(['Month', 'CancellationCode']).size()
Cancel_Month

Month  CancellationCode
1      N                   183527
2      N                   189534
3      N                   200842
4      N                   155264
5      N                   153072
6      N                   200914
7      N                   182945
8      N                   162648
9      N                    95539
10     A                       33
       B                       17
       C                        9
       N                   103466
11     A                       41
       B                       40
       C                       13
       N                   105469
12     A                      172
       B                      250
       C                       58
       N                   202905
dtype: int64

Pel que fa a les cancel·lacions, en agrupar-les per mesos,  
estan concentrades a final d'any, als mesos d'octubre, novembre i desembre.

In [13]:
data.isna().sum()

Month                   0
DayofMonth              0
DayOfWeek               0
UniqueCarrier           0
ActualElapsedTime    8387
CRSElapsedTime        198
AirTime              8387
ArrDelay             8387
DepDelay                0
Origin                  0
Dest                    0
Distance                0
TaxiIn               7110
TaxiOut               455
Cancelled               0
CancellationCode        0
Diverted                0
CarrierDelay            0
WeatherDelay            0
NASDelay                0
SecurityDelay           0
LateAircraftDelay       0
dtype: int64

Hi ha diverses columnes on falta força informació rellevant, com el temps real transcorregut, de manera que en aquestes columnes, la informació no és fiable, per la qual cosa he optat per suprimir algunes. El dataset és força gran, així que no hauria de falsejar la informació general.

In [14]:
SinNull = data.loc[(-data['ActualElapsedTime'].isnull()) & (-data['AirTime'].isnull())]
SinNull.isna().sum()

Month                0
DayofMonth           0
DayOfWeek            0
UniqueCarrier        0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
TaxiIn               0
TaxiOut              0
Cancelled            0
CancellationCode     0
Diverted             0
CarrierDelay         0
WeatherDelay         0
NASDelay             0
SecurityDelay        0
LateAircraftDelay    0
dtype: int64

Treient les files on els valors de *ActualElapsedTime* i també de *AirTime* són nuls, el dataset queda net.

In [15]:
SinNull = SinNull.copy()
SinNull['Gap'] = (SinNull['CRSElapsedTime']) - (SinNull['ActualElapsedTime'])

In [16]:
SinNull['AvSpeed'] = round((SinNull['Distance']) / ((SinNull['AirTime']) / 60),2)

In [17]:
Dist_Gap = 4000
SinNull['DistCode'] = ' '
ListDistCode = []

for i in SinNull['Distance']:
    if i <= Dist_Gap:
        ListDistCode.append('Short') 
    elif i > Dist_Gap:
        ListDistCode.append('Long')
    else:
        ListDistCode.append('NaN')
SinNull['DistCode'] = ListDistCode

In [18]:
SinNull['Status'] = ' '
ListStatus = []

for i in SinNull['Gap']:
    if i <= 0:
        ListStatus.append('On Time') 
    else :
        ListStatus.append('Delayed')
SinNull['Status'] = ListStatus

In [19]:
SinNull.head(2)

Month  DayofMonth  DayOfWeek UniqueCarrier  ActualElapsedTime  \
0      1           3          4            WN              128.0   
1      1           3          4            WN              128.0   

   CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  \
0           150.0    116.0     -14.0       8.0    IAD  TPA       810     4.0   
1           145.0    113.0       2.0      19.0    IAD  TPA       810     5.0   

   TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  WeatherDelay  \
0      8.0          0                N         0           0.0           0.0   
1     10.0          0                N         0           0.0           0.0   

   NASDelay  SecurityDelay  LateAircraftDelay   Gap  AvSpeed DistCode   Status  
0       0.0            0.0                0.0  22.0   418.97    Short  Delayed  
1       0.0            0.0                0.0  17.0   430.09    Short  Delayed

He afegit algunes columnes que crec que poden ser interessants,
bé per ajudar a visualitzar el dataframe o bé per poder treballar-hi més endavant:
- La columna *Gap* És la diferència entre el temps real i l'estimat.
- La columna *DistCode* Divideix els vols entre Llarga i Curta diastància, on es veu que el volum de vols de llarga distància és molt petit en comparació amb els de curta, cosa que sembla que connecta amb el filtratge anterior on la majoria de retards es donaven als vols de curta distància.
- La columna *AvSpeed* És la velocitat mitjana de l'avió (A l'aire) i està expressada a Millas/Hora.
- La columna *Status* mostra si el vol s'ha endarrerit o no.

In [20]:
SinNull.groupby(['DistCode']).size()

DistCode
Long        1198
Short    1927173
dtype: int64

In [21]:
AeroDelay = dict(SinNull[SinNull['Gap'] == True].groupby('UniqueCarrier').size())

Sorted_AeroDelay = dict(sorted(AeroDelay.items(), key=lambda item: item[1], reverse=True))
Sorted_AeroDelay

{'WN': 15744,
 'AA': 5978,
 'MQ': 5333,
 'OO': 5332,
 'UA': 3955,
 'DL': 3490,
 'XE': 3324,
 'EV': 2815,
 'FL': 2623,
 'US': 2508,
 'NW': 2478,
 'CO': 2455,
 'YV': 2316,
 '9E': 1587,
 'OH': 1579,
 'AS': 1444,
 'B6': 1358,
 'F9': 1212,
 'HA': 635,
 'AQ': 53}

A dalt es mostren els retards acumulats en funció de la companyia. Estan ordenats de més a menys retard.

In [22]:
Sort_Dist = SinNull.sort_values('Distance', ascending=False)
Sort_Dist.loc[:,['Distance', 'Origin', 'Dest']]

Distance Origin Dest
362407       4962    EWR  HNL
556368       4962    EWR  HNL
556375       4962    EWR  HNL
556374       4962    EWR  HNL
556373       4962    EWR  HNL
...           ...    ...  ...
169785         30    JFK  HPN
47146          24    IAD  DCA
521905         24    IAH  HOU
1310766        11    JFK  LGA
774654         11    JFK  LGA

[1928371 rows x 3 columns]

A dalt es mostra una taula amb les rutes ordenades de major a menor distància, i sota es mostren les
rutes que acumulen més retards.

In [23]:
Sort_Retr = SinNull.sort_values('Gap', ascending=False)
Sort_Retr.loc[:,['Gap', 'Origin', 'Dest']].head(5)

Gap Origin Dest
776886   181.0    SLC  DEN
545038   151.0    LGA  MIA
522132   142.0    PIT  ATL
1110539  128.0    AUS  SAN
919798   125.0    SAN  ABQ

In [24]:
Actual = len(SinNull)
print(Actual)
Nueva = 200000
print(Nueva)
import random
Indices_Nueva = random.sample(range(Actual + 1), Nueva)
Data_Reducido = SinNull.iloc[Indices_Nueva]
df_200 = Data_Reducido.copy()
df_200

1928371
200000


Month  DayofMonth  DayOfWeek UniqueCarrier  ActualElapsedTime  \
743584       5          17          6            WN              120.0   
735926       5          11          7            WN               64.0   
953450       6          25          3            OO               75.0   
1028991      6           5          4            NW              261.0   
151393       1          23          3            AA              134.0   
...        ...         ...        ...           ...                ...   
1569399     10          15          3            UA              220.0   
1036185      6          28          6            AA              407.0   
1771638     12          30          2            WN              345.0   
475937       3          27          4            DL              106.0   
302400       2          28          4            FL              100.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
743584            130.0    106.0      16.0      26.0    PDX  LAS       762   
735926             60.0     46.0      12.0       8.0    LAS  BUR       223   
953450             75.0     57.0      29.0      29.0    MKE  CMH       331   
1028991           260.0    231.0      25.0      24.0    DTW  LAS      1750   
151393            140.0    112.0      19.0      25.0    DFW  MSP       852   
...                 ...      ...       ...       ...    ...  ...       ...   
1569399           232.0    205.0      14.0      26.0    ORD  PHX      1440   
1036185           340.0    303.0      86.0      19.0    LAX  JFK      2475   
1771638           345.0    324.0      27.0      27.0    PHL  LAS      2176   
475937            114.0     85.0       9.0      17.0    BWI  ATL       576   
302400            100.0     79.0       7.0       7.0    ATL  CAK       528   

         TaxiIn  TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  \
743584      6.0      8.0          0                N         0           3.0   
735926      2.0     16.0          0                N         0           0.0   
953450      5.0     13.0          0                N         0           7.0   
1028991     7.0     23.0          0                N         0          24.0   
151393      8.0     14.0          0                N         0           9.0   
...         ...      ...        ...              ...       ...           ...   
1569399     3.0     12.0          0                N         0           0.0   
1036185    88.0     16.0          0                N         0           0.0   
1771638     4.0     17.0          0                N         0          27.0   
475937      8.0     13.0          0                N         0           0.0   
302400      5.0     16.0          0                N         0           0.0   

         WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay   Gap  \
743584            0.0       0.0            0.0               13.0  10.0   
735926            0.0       0.0            0.0                0.0  -4.0   
953450            0.0       0.0            0.0               22.0   0.0   
1028991           0.0       1.0            0.0                0.0  -1.0   
151393            0.0       0.0            0.0               10.0   6.0   
...               ...       ...            ...                ...   ...   
1569399           0.0       0.0            0.0                0.0  12.0   
1036185           0.0      86.0            0.0                0.0 -67.0   
1771638           0.0       0.0            0.0                0.0   0.0   
475937            0.0       0.0            0.0                0.0   8.0   
302400            0.0       0.0            0.0                0.0   0.0   

         AvSpeed DistCode   Status  
743584    431.32    Short  Delayed  
735926    290.87    Short  On Time  
953450    348.42    Short  On Time  
1028991   454.55    Short  On Time  
151393    456.43    Short  Delayed  
...          ...      ...      ...  
1569399   421.46    Short  Delayed  
1036185   490.10    Short

In [25]:
df_200_ord = df_200.sort_index()
df_200_ord = df_200_ord.copy()
df_200_ord.head()

Month  DayofMonth  DayOfWeek UniqueCarrier  ActualElapsedTime  \
14      1           3          4            WN               56.0   
25      1           3          4            WN              127.0   
49      1           3          4            WN               75.0   
55      1           3          4            WN              148.0   
66      1           3          4            WN               64.0   

    CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  \
14            70.0     49.0       6.0      20.0    ISP  BWI       220     2.0   
25           145.0    113.0      11.0      29.0    ISP  MDW       765     8.0   
49            85.0     60.0      43.0      53.0    LAS  ABQ       487     4.0   
55           160.0    131.0       4.0      16.0    LAS  AUS      1090     3.0   
66            60.0     48.0      55.0      51.0    LAS  BUR       223     3.0   

    TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  WeatherDelay  \
14      5.0          0                N         0           0.0           0.0   
25      6.0          0                N         0           0.0           0.0   
49     11.0          0                N         0          15.0           0.0   
55     14.0          0                N         0           0.0           0.0   
66     13.0          0                N         0           4.0           0.0   

    NASDelay  SecurityDelay  LateAircraftDelay   Gap  AvSpeed DistCode  \
14       0.0            0.0                0.0  14.0   269.39    Short   
25       0.0            0.0                0.0  18.0   406.19    Short   
49       0.0            0.0               28.0  10.0   487.00    Short   
55       0.0            0.0                0.0  12.0   499.24    Short   
66       4.0            0.0               47.0  -4.0   278.75    Short   

     Status  
14  Delayed  
25  Delayed  
49  Delayed  
55  Delayed  
66  On Time

In [26]:
df_200_ord.to_excel('df_200_ord.xlsx', index = False)